<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="robo_diferenciado">

    <xacro:macro name="box_inertia" params="m x y z">
        <inertial>
            <mass value="${m}"/>
            <inertia ixx="${(m/12) * (y*y + z*z)}" ixy="0" ixz="0"
                     iyy="${(m/12) * (x*x + z*z)}" iyz="0"
                     izz="${(m/12) * (x*x + y*y)}"/>
        </inertial>
    </xacro:macro>

    <xacro:macro name="cylinder_inertia" params="m r h">
        <inertial>
            <origin xyz="0 0 0" rpy="${pi/2} 0 0" />
            <mass value="${m}"/>
            <inertia ixx="${(m/12) * (3*r*r + h*h)}" ixy="0" ixz="0"
                     iyy="${(m/12) * (3*r*r + h*h)}" iyz="0"
                     izz="${(m/2) * (r*r)}"/>
        </inertial>
    </xacro:macro>

    <xacro:macro name="sphere_inertia" params="m r">
        <inertial>
            <mass value="${m}"/>
            <inertia ixx="${(2/5) * m * (r*r)}" ixy="0.0" ixz="0.0"
                     iyy="${(2/5) * m * (r*r)}" iyz="0.0"
                     izz="${(2/5) * m * (r*r)}"/>
        </inertial>
    </xacro:macro>

    <xacro:property name="chassis_mass" value="5.0"/>
    <xacro:property name="chassis_length" value="0.4"/>
    <xacro:property name="chassis_width" value="0.3"/>
    <xacro:property name="chassis_height" value="0.1"/>

    <xacro:property name="wheel_radius" value="0.1"/>
    <xacro:property name="wheel_width" value="0.05"/>
    <xacro:property name="wheel_mass" value="1.0"/>
    <xacro:property name="wheel_offset_y" value="0.175"/> <xacro:property name="caster_radius" value="0.05"/>
    <xacro:property name="caster_mass" value="0.5"/>
    <xacro:property name="caster_offset_x" value="0.15"/>

    <link name="base_link">
    </link>

    <link name="chassis">
        <visual>
            <origin xyz="0 0 ${chassis_height/2}" rpy="0 0 0"/>
            <geometry>
                <box size="${chassis_length} ${chassis_width} ${chassis_height}"/>
            </geometry>
            <material name="blue">
                <color rgba="0 0 1 1"/>
            </material>
        </visual>
        <collision>
            <origin xyz="0 0 ${chassis_height/2}" rpy="0 0 0"/>
            <geometry>
                <box size="${chassis_length} ${chassis_width} ${chassis_height}"/>
            </geometry>
        </collision>
        <xacro:box_inertia m="${chassis_mass}" x="${chassis_length}" y="${chassis_width}" z="${chassis_height}"/>
    </link>

    <joint name="base_joint" type="fixed">
        <parent link="base_link"/>
        <child link="chassis"/>
        <origin xyz="0 0 ${wheel_radius}" rpy="0 0 0"/>
    </joint>

    <link name="left_wheel">
        <visual>
            <origin xyz="0 0 0" rpy="${pi/2} 0 0"/>
            <geometry>
                <cylinder radius="${wheel_radius}" length="${wheel_width}"/>
            </geometry>
            <material name="black">
                <color rgba="0 0 0 1"/>
            </material>
        </visual>
        <collision>
            <origin xyz="0 0 0" rpy="${pi/2} 0 0"/>
            <geometry>
                <cylinder radius="${wheel_radius}" length="${wheel_width}"/>
            </geometry>
        </collision>
        <xacro:cylinder_inertia m="${wheel_mass}" r="${wheel_radius}" h="${wheel_width}"/>
    </link>

    <joint name="left_wheel_joint" type="continuous">
        <parent link="chassis"/>
        <child link="left_wheel"/>
        <origin xyz="-0.1 ${wheel_offset_y} 0" rpy="0 0 0"/> <axis xyz="0 1 0"/>
    </joint>

    <link name="right_wheel">
        <visual>
            <origin xyz="0 0 0" rpy="${pi/2} 0 0"/>
            <geometry>
                <cylinder radius="${wheel_radius}" length="${wheel_width}"/>
            </geometry>
            <material name="black">
                <color rgba="0 0 0 1"/>
            </material>
        </visual>
        <collision>
            <origin xyz="0 0 0" rpy="${pi/2} 0 0"/>
            <geometry>
                <cylinder radius="${wheel_radius}" length="${wheel_width}"/>
            </geometry>
        </collision>
        <xacro:cylinder_inertia m="${wheel_mass}" r="${wheel_radius}" h="${wheel_width}"/>
    </link>

    <joint name="right_wheel_joint" type="continuous">
        <parent link="chassis"/>
        <child link="right_wheel"/>
        <origin xyz="-0.1 -${wheel_offset_y} 0" rpy="0 0 0"/> <axis xyz="0 1 0"/>
    </joint>

    <link name="caster_wheel">
        <visual>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <geometry>
                <sphere radius="${caster_radius}"/>
            </geometry>
            <material name="white">
                <color rgba="1 1 1 1"/>
            </material>
        </visual>
        <collision>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <geometry>
                <sphere radius="${caster_radius}"/>
            </geometry>
        </collision>
        <xacro:sphere_inertia m="${caster_mass}" r="${caster_radius}"/>
    </link>

    <joint name="caster_joint" type="fixed">
        <parent link="chassis"/>
        <child link="caster_wheel"/>
        <origin xyz="${caster_offset_x} 0 -${wheel_radius - caster_radius}" rpy="0 0 0"/>
    </joint>

    <gazebo reference="left_wheel">
        <mu1>1.0</mu1> <mu2>1.0</mu2>
        <material>Gazebo/Black</material>
    </gazebo>
    <gazebo reference="right_wheel">
        <mu1>1.0</mu1> <mu2>1.0</mu2>
        <material>Gazebo/Black</material>
    </gazebo>

    <gazebo reference="caster_wheel">
        <mu1>0.001</mu1> <mu2>0.001</mu2>
        <material>Gazebo/White</material>
    </gazebo>

</robot>
